File: demo-iys-rest-scraping.java/src/main/java/demo/file/TsvFile.java
package demo.file;

import demo.json.TargetsListFile;
import demo.model.Target;
import demo.model.TargetsList;

import java.util.Comparator;
import java.util.function.Function;
import java.util.stream.Collectors;

public class TsvFile {

    public static void main(String[] args) {
        final String fileName = "reader/skills.json";
        TargetsListFile targetsListFile = new TargetsListFile(fileName);
        TargetsList targetsList = targetsListFile.load();

        saveSortById(targetsList, fileName);
        saveSortByName(targetsList, fileName);
    }

    private static void saveSortById(TargetsList targetsList, String fileName) {
        save(
                targetsList, fileName, ".sort by id.tsv",
                (o1, o2) -> o1.getId().compareTo(o2.getId()),
                (o) -> o.getId() + '\t' + o.getName()
        );
    }

    private static void saveSortByName(TargetsList targetsList, String fileName) {
        save(
                targetsList, fileName, ".sort by name.tsv",
                (o1, o2) -> o1.getName().compareTo(o2.getName()),
                (o) -> o.getName() + '\t' + o.getId()
        );
    }

    private static void save(TargetsList targetsList, String fileName, String extension, Comparator<Target> comparator, Function<Target, String> mapper) {
        String value = targetsList.getTargets()
                .stream()
                .sorted(comparator)
                .map(mapper)
                .collect(Collectors.joining("\n"));

        TextFile textFile = new TextFile(fileName.replace(".json", extension));
        textFile.save(value);
    }
}

File: demo-iys-rest-scraping.kotlin/src/main/kotlin/demo/file/TsvFile.kt
package demo.file

import demo.json.TargetsListFile
import demo.model.Target
import demo.model.TargetsList
import java.util.*
import java.util.function.Function

object TsvFile {

    @JvmStatic fun main(args: Array<String>) {
        val fileName = "reader/skills.json"
        val targetsListFile = TargetsListFile(fileName)
        val targetsList = targetsListFile.load()

        saveSortById(targetsList, fileName)
        saveSortByName(targetsList, fileName)
    }

    private fun saveSortById(targetsList: TargetsList, fileName: String) {
        save(
                targetsList, fileName, ".sort by id.tsv",
                Comparator<Target> { o1, o2 -> o1.id.compareTo(o2.id) },
                Function<Target, String> { o -> o.id + '\t' + o.name }
        )
    }

    private fun saveSortByName(targetsList: TargetsList, fileName: String) {
        save(
                targetsList, fileName, ".sort by name.tsv",
                Comparator<Target> { o1, o2 -> o1.name.compareTo(o2.name) },
                Function<Target, String> { o -> o.name + '\t' + o.id }
        )
    }

    private fun save(targetsList: TargetsList, fileName: String, extension: String, comparator: Comparator<Target>, mapper: Function<Target, String>) {
        val value: List<String> = targetsList.targets
                .sortedWith(comparator)
                .map { target -> mapper.apply(target) }

        val textFile = TextFile(fileName.replace(".json", extension))
        textFile.save(value)
    }
}
